### PR TITLE
Adding of clone_schema() macro

### PR DESCRIPTION
**Pull request by oleg-kolesnikov**
## What is this? 
_This PR adds `clone_schema()`, which clones tables, sequences and views frome one schema to another. The range of objects to be copied by can be limited via `comment_tag` argument. The macro is supposed as a part of potential Blue/Green Deploy of dbt projects._

## What changes in this PR? 
* _this adds new macro `clone_schema()` and corresponding models and tests to it._

## How does this impact our users?
* _this provides new options of data management might be useful for organisation of Blue/Green Deploy of dbt projects._

## PR Quality
- [ + ] does `docker-compose exec testxdb test` run successfully?
- [ + ] does `docker-compose exec testxdb docs` build docs without errors?
- [ + ] does `docker-compose exec testxdb coverage` pass coverage standards?
- [ - ] did you leave the codebase better than you found it? (I'm not sure%. This time too sophisticated code happened.)

@Health-Union/hu-data make sure to include a version tag in the merge commit!

### CODE REVIEW

File: macros/clone_schema.sql [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+       SUPPORTS:
+            - Postgres
+            - Snowflake
+    */#}
+
+    {%- if target.type == 'postgres' -%}
+
+        {% set set_pg_function %}
+
+            CREATE OR REPLACE FUNCTION clone_schema(source_schema text,
```
Comment (oleg-kolesnikov) [inline macros/clone_schema.sql:18]: I've found that there are neither special clone schema command or procedure in Postgres. Finally I've decided to make this job via single crate/alter commands that are formed while iteration on lists of original objects in loops. This part of macro is based on [this script](https://www.postgresql.org/message-id/CANu8FiyJtt-0q%3DbkUxyra66tHi6FFzgU8TqVR2aahseCBDDntA%40mail.gmail.com) from Internet.

File: macros/clone_schema.sql [hunk 2]
```diff
@@ -0,0 +1,266 @@
+{%- macro clone_schema(schema_one, schema_two, comment_tag='') -%}
```
Comment (oleg-kolesnikov) [inline macros/clone_schema.sql:1]: I've found that Postgres doesn't support tags mechanism as [Snowflake](https://docs.snowflake.com/en/user-guide/object-tagging.html) does so to make this macro more general I decided to use metadata field `comment` to filter objects we need to have.
I believe that only we would need to do for using this macro after this PR are:
1. Mark incremental tables and sequences in `dwh` by special comment 'incremental'.
2. Use command `dbt run-operation clone_schema --args '{schema_one: prod, schema_two: stage, comment_tag: incremental}' --target ANALYSIS`

File: macros/clone_schema.sql [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+    {#/* Copies tables, sequences and views from `schema_one` to `schema_two` if `comment_tag` isn't specified. If `comment_tag` argument is specified, it copies only tables and sequences that have `comment` metadata field equal to the passed value of `comment_tag` argument.
```
Comment (oleg-kolesnikov) [inline macros/clone_schema.sql:3]: To be honest description of this macro doesn't include the fact that not passing any value to `comment_tag` arg makes it copy not only tables, sequences and views but functions too. I'll extend it by this details.

File: macros/clone_schema.sql [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+            $BODY$
+            LANGUAGE plpgsql VOLATILE
+            COST 100;
+
+        {% endset %} 
+
+        {% do run_query(set_pg_function) %}
+
+        {% set sql %}
+            SELECT clone_schema('{{schema_one}}', '{{schema_two}}', '{{comment_tag}}', true);
```
Comment (oleg-kolesnikov) [inline macros/clone_schema.sql:217]: Here we are setting SQL-statement calling to  clone_schema Postgres function declared before.

File: macros/clone_schema.sql [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+                {% endfor %}
+            {% endset %}
+
+        {%- endif -%}
+
+    {%- else -%}
+        {{ xdb.not_supported_exception('The clone_schema() macro doesn`t support this type of database target.') }}
+    {%- endif -%}
+
+    {% do run_query(sql) %}
```
Comment (oleg-kolesnikov) [inline macros/clone_schema.sql:258]: Here we are running `SQL`-statements block doing all job we need.

File: macros/clone_schema.sql (hunk 1 above)
Comment (oleg-kolesnikov) [reply macros/clone_schema.sql:18]: Code style of this part are questionable and might not clear - I've left a lot of original artefacts and mechanisms here. I can rewrite it more clearly but suppose such sort of work a bit redundant for now. At least, according to tests it works%)

File: macros/clone_schema.sql [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+            --2. Create sequences.
+            
+            SELECT count(*) INTO count_rows 
+            FROM information_schema.sequences as t
+            LEFT JOIN pg_catalog.pg_description AS pgc
+            ON CONCAT_WS('.', t.sequence_schema, t.sequence_name)::regclass::oid = pgc.objoid
+            WHERE t.sequence_schema = quote_ident(source_schema)
+                AND case when comment_tag = '' then '' else description end = comment_tag;
+
+            IF count_rows > 0 THEN
```
Comment (oleg-kolesnikov) [inline macros/clone_schema.sql:64]: Check if original schema has sequences.

File: macros/clone_schema.sql [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+                FOR object in
+                    SELECT sequence_name::text
```
Comment (oleg-kolesnikov) [inline macros/clone_schema.sql:66]: Getting of sequences list.

File: macros/clone_schema.sql [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+                    FROM information_schema.sequences as t
+                    LEFT JOIN pg_catalog.pg_description AS pgc
+                    ON CONCAT_WS('.', sequence_schema, sequence_name)::regclass::oid = pgc.objoid
+                    WHERE t.sequence_schema = quote_ident(source_schema)
+                        AND case when comment_tag = '' then '' else pgc.description end = comment_tag
+                LOOP
+                    EXECUTE 'CREATE SEQUENCE ' || quote_ident(dest_schema) || '.' || quote_ident(object);
```
Comment (oleg-kolesnikov) [inline macros/clone_schema.sql:73]: Creation set of commands in the loop.

File: macros/clone_schema.sql [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+            SELECT count(*) INTO count_rows 
+            FROM information_schema.tables as t
+            LEFT JOIN pg_catalog.pg_description AS pgc
+            ON CONCAT_WS('.', table_schema, table_name)::regclass::oid = pgc.objoid
+            WHERE table_schema = quote_ident(source_schema)
+                AND table_type = 'BASE TABLE'
+                AND case when comment_tag = '' then '' else description end = comment_tag;
+
+            IF count_rows > 0 THEN
```
Comment (oleg-kolesnikov) [inline macros/clone_schema.sql:115]: See explanation comments for sequences above.

File: macros/clone_schema.sql [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+        {% do run_query(set_pg_function) %}
+
+        {% set sql %}
+            SELECT clone_schema('{{schema_one}}', '{{schema_two}}', '{{comment_tag}}', true);
+        {% endset %}
+
+    {%- elif target.type == 'snowflake' -%}
+        {%- if comment_tag == '' -%}
+            {% set sql %}
+                CREATE OR REPLACE SCHEMA {{schema_two}} CLONE {{schema_one}};
```
Comment (oleg-kolesnikov) [inline macros/clone_schema.sql:223]: And Snowflake does almost the same work as this frightening Postgres function but by a single command%)

File: test_xdb/models/under_test/clone_schema_test.sql [hunk 11]
```diff
@@ -0,0 +1,200 @@
+{{config({
```
Comment (oleg-kolesnikov) [inline test_xdb/models/under_test/clone_schema_test.sql:1]: This model shows us how many deltas in tables and views counts in test schemas we have.

File: test_xdb/models/under_test/clone_schema_test.sql [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+    "tags":["exclude_bigquery", "exclude_bigquery_tests"],
+    "pre-hook": [{"sql": "DROP SCHEMA IF EXISTS clone_schema_one CASCADE;
```
Comment (oleg-kolesnikov) [inline test_xdb/models/under_test/clone_schema_test.sql:3]: Here we are specifying test schemas and performing `clone_schema()` macro on them.

File: test_xdb/models/under_test/clone_schema_tables_metadata_equality_test.sql [hunk 13]
```diff
@@ -0,0 +1,135 @@
+--depends_on: {{ ref('clone_schema_tables_data_equality_test') }}
```
Comment (oleg-kolesnikov) [inline test_xdb/models/under_test/clone_schema_tables_metadata_equality_test.sql:1]: This model shows us lists of metadata fields with deltas between original and copied tables and views in test schemas.

File: test_xdb/models/under_test/clone_schema_tables_data_equality_test.sql [hunk 14]
```diff
@@ -0,0 +1,89 @@
+--depends_on: {{ ref('clone_schema_test') }}
```
Comment (oleg-kolesnikov) [inline test_xdb/models/under_test/clone_schema_tables_data_equality_test.sql:1]: This model shows us counts of not matched rows between original and copied tables/views in test schemas.

File: test_xdb/models/under_test/clone_schema_sequences_metadata_equality_test.sql [hunk 15]
```diff
@@ -0,0 +1,129 @@
+--depends_on: {{ ref('clone_schema_tables_metadata_equality_test') }}
```
Comment (oleg-kolesnikov) [inline test_xdb/models/under_test/clone_schema_sequences_metadata_equality_test.sql:1]: This model shows us lists of metadata fields with deltas between original and copied sequences in test schemas.

File: macros/clone_schema.sql [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+            sq_is_cycled boolean;
+            sq_cycled char(10);
+
+            BEGIN
+            --1. Create schema.
+            EXECUTE 'DROP SCHEMA IF EXISTS ' || quote_ident(dest_schema) || ' CASCADE; CREATE SCHEMA ' || quote_ident(dest_schema);
+
+            --2. Create sequences.
+            
+            SELECT count(*) INTO count_rows
```
Comment (oleg-kolesnikov) [inline macros/clone_schema.sql:57]: Fetching of sequences counts.

File: macros/clone_schema.sql (hunk 3 above)
Comment (oleg-kolesnikov) [reply macros/clone_schema.sql:3]: I haven't covered this `functions` behaviour by tests and consider it as a redundant work for now. We can skip them on Postgres target easily but I still don't check how Snowflake deal with them via CLONE SCHEMA command.

**Review by philribbens — APPROVED** [review]
This looks good to me. The only thing I would suggest is including the comments in the actual code itself, not just in the PR comments. That way they outlast the PR and can be seen in the code directly by us in the future.

File: macros/clone_schema.sql (hunk 16 above)
Comment (philribbens) [reply macros/clone_schema.sql:69]: @oleg-kolesnikov Could you include this (and the other comments you made on the PR) as inline comments in the code?

Comment (oleg-kolesnikov) [reply macros/clone_schema.sql:70]: Seems like have done it.

**Pull request merged**